documentation/interrupt_handling.md

```diff
@@ -1,23 +1,79 @@
-Entry and exit of secure world
-==============================
-
-#Introduction
-Depending on configuration of the system secure world can be entered during
-different conditions. This document will describe only the configuration
-used by OP-TEE.
-
-Monitor vector is VBAR_EL3 in AArch64 and MVBAR in ARMv7/AArch32. State
-vector is any of:
-* VBAR_EL2 or VBAR_EL1 (secure and non-secure) for AArch64, depending on
-  configuration of hypervisor
-* HVBAR or VBAR (secure and non-secure) for ARMv7, depending on
-  configuration of hypervisor
-
-The processor is configured to use:
-* Monitor vector for FIQ exceptions while SCR_NS is set and state vector
-  when SCR_NS is cleared
-* Monitor vector for SMC exceptions
-* State vector for IRQ exceptions
+Exceptions, Interrupts and World Contextes
+==========================================
+
+This document describes describes how optee_os handles switches of
+world execution context based on SMC exceptions and interrupt notifications
+Interrupt notifications are IRQ/FIQ exceptions which may also imply switching
+of world execution context: non secure to secure, or secure to non secure.
+
+## Content
+
+1. [Context Switches and Exceptions](#1-context-switches-and-exceptions)
+2. [Normal World Invokes the optee_os Through a SMC](#2-normal-world-invokes-the-optee_os-through-a-smc)
+3. [Deliver a Non Secure Interrupt to the Non Secure World](#3-deliver-a-non-secure-interrupt-to-the-non-secure-world)
+4. [Deliver a Secure Interrupt to the Secure World](#4-deliver-a-secure-interrupt-to-the-secure-world)
+5. [TEE Thread Scheduling](#5-tee-thread-scheduling)
+
+# 1. Context Switches and Exceptions
+
+## 1.1. Usecases of world context switch
+
+When the normal world invokes the secure world, the normal world executes
+a SMC instruction. The SMC exception is always trapped by the Monitor. If
+the related service targets the trusted OS, the monitor invokes optee_os.
```

Comment:
IMHO, it would be really nice to have some links to actual code (for example, "for additional details about handling smc in optee_os core check `tee_entry_std in `core/arch/arm/tee/entry_std.c``"). I've noticed that it's really for people that are not familiar with code base make a connection between what is written here and actual parts of code, that implements described behavior

Reply:
This is the introduction, just an overview so not much details.